vpr/src/analytical_place/full_legalizer.cpp

```diff
@@ -281,6 +291,710 @@ static LegalizationClusterId create_new_cluster(PackMoleculeId seed_molecule_id,
+        for (LegalizationClusterId cluster_id: created_clusters) {
+            if (!cluster_legalizer.check_cluster_legality(cluster_id)) {
+                for (PackMoleculeId mol_id : cluster_legalizer.get_cluster_molecules(cluster_id)) {
+                    illegal_cluster_mols.push_back(mol_id);
+                }
+                // Erase related data of illegal cluster
+                loc_to_cluster_id_placed.erase(cluster_locs[cluster_id]);
+                cluster_legalizer.destroy_cluster(cluster_id);
+                tile_clusters_matrix[tile_loc.layer_num][tile_loc.x][tile_loc.y].erase(cluster_id);
+            } else {
+                cluster_legalizer.clean_cluster(cluster_id);
+            }
+        }
+
+        // If there are any illegal molecules, set the legalization strategy to
+        // full and try to cluster the unclustered molecules in same tile again.
+        if (!illegal_cluster_mols.empty()) {
+            cluster_legalizer.set_legalization_strategy(ClusterLegalizationStrategy::FULL);
+            created_clusters.clear();
+            cluster_molecules_in_tile(tile_loc,
+                                    tile_type,
+                                    illegal_cluster_mols,
+                                    cluster_legalizer,
+                                    primitive_candidate_block_types,
+                                    created_clusters);
+
+            // Clean clusters created with full strategy not to increase memory footprint.
+            for (LegalizationClusterId cluster_id : created_clusters) {
+                cluster_legalizer.clean_cluster(cluster_id);
+            }
+        }
+    }
+}
+
+void FlatRecon::neighbor_clustering(ClusterLegalizer& cluster_legalizer,
+                                    const vtr::vector<LogicalModelId, std::vector<t_logical_block_type_ptr>>& primitive_candidate_block_types,
+                                    std::unordered_set<PackMoleculeId>& mols_clustered) {
+    vtr::ScopedStartFinishTimer neigh_pass_clustering("Neighbor Pass Clustering");
+    
+    // Iterate over molecules and try to join the unclustered ones to their
+    // already created 8-neighboring tile clusters.
+    for (APBlockId blk_id : ap_netlist_.blocks()) {
+        // Get unclustered block and its location.
+        PackMoleculeId molecule_id = ap_netlist_.block_molecule(blk_id);
+        t_physical_tile_loc loc = mol_desired_physical_tile_loc[molecule_id];
+
+        // Skip the already clustered molecules.
+        if (cluster_legalizer.is_mol_clustered(molecule_id))
+            continue;
+
+        // Skip the io blocks in that pass. This aims to avoid unneccessary recreation of clusters neighboring io blocks.
+        // TODO: This might be handled more generally instead of checking block name.
+        t_logical_block_type_ptr block_type = infer_molecule_logical_block_type(molecule_id, prepacker_, primitive_candidate_block_types);
```

Comment:
This is pretty hacky.  Why are we skipping the IO blocks?
Use the is_io function instead: https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/8c48d501b2586a81bfd1abc4f20ff9bbc42b0b22/libs/libarchfpga/src/physical_types.cpp#L232
Also seems like this is really trying to avoid checking irrelevant parts of the grid, which would be better to do using the compressed grid or some general code to see if you are at the grid edge.